nengo_gui/examples/tutorial/25-spa-parse.py

```diff
@@ -4,53 +4,67 @@
+# with the semantic pointer NOUN and the verb stored in the verb memory binds
+# with the semantic pointer VERB in the phrase state.  This binding is
+# implemented with cortical actions. Notice that the actions can looks for
+# multiple different things at the same time by using the + operation.
+# Overall, this shows how we can use a single input modality (vision) and treat
+# the information in different ways as appropriate.
+
 # The last two actions deal with recognizing and executing actions.  The
 # first one looks for the VERB WRITE, and if it sees this it will take
-# whatever is in memory, extract out the NOUN, and send that to the hand.
+# whatever is in phrase, extract out the NOUN, and send that to the hand.
 # Furthermore, it will only do this if it doesn't currently see anything in
 # vision (via the subtraction).  This is to make sure it waits until it is
 # getting no more visual input before responding.
 
-# To test the model, try presenting WRITE to the vision Buffer.  The memory
-# should fill with WRITE*VERB.  Now change the vision input to B.  The memory
-# should now fill with NOUN*B.  If you get rid of the visual input (by setting
-# the value to nothing), it shuold sent B to the hand.  If you try the same
+# To test the model, try presenting WRITE to the vision State.  The phrase
+# should fill with WRITE*VERB.  Now change the vision input to HI.  The phrase
+# should now fill with NOUN*HI.  If you get rid of the visual input (by setting
+# the value to nothing), it shuold sent HI to the hand.  If you try the same
 # thing with SAY it should send the result to speech.
 
 import nengo
 import nengo.spa as spa
 
-D = 32  # the dimensionality of the vectors
+D = 64  # the dimensionality of the vectors
 
 model = spa.SPA()
 with model:
-    model.vision = spa.Buffer(D)
-    
-    model.memory = spa.Memory(D, synapse=0.2)
-    model.speech = spa.Buffer(D)
-    model.hand = spa.Buffer(D)
+    # perceptual input
+    model.vision = spa.State(D)
+
+    # linguistic components
+    model.phrase = spa.State(D)
+    model.verb = spa.State(D, feedback=1, feedback_synapse=0.01)
+    model.noun = spa.State(D, feedback=1, feedback_synapse=0.01)
+
+    # motor components
+    model.speech = spa.State(D)
+    model.hand = spa.State(D)
 
     actions = spa.Actions(
-        'dot(vision, WRITE+SAY) --> memory=vision*VERB',
-        'dot(vision, A+B+C+D+E) --> memory=vision*NOUN',
-        'dot(memory, VERB*WRITE) - dot(vision, WRITE+SAY+A+B+C+D+E)'
-            '--> hand=memory*~NOUN',
-        'dot(memory, VERB*SAY) - dot(vision, WRITE+SAY+A+B+C+D+E)'
-            '--> speech=memory*~NOUN',
+        'dot(vision, WRITE+SAY) --> verb=vision',
+        'dot(vision, YES+NO+HI+BYE+OK) --> noun=vision',
+        'dot(phrase, VERB*WRITE) - dot(vision, WRITE+SAY+YES+NO+HI+BYE+OK)'
+            '--> hand=phrase*~NOUN',
+        'dot(phrase, VERB*SAY) - dot(vision, WRITE+SAY+YES+NO+HI+BYE+OK)'
+            '--> speech=phrase*~NOUN',
```

Comment:
I still find this model to sometimes fail for me.  Looking at the thalamus and bg plots, it looks like it's sometimes selecting the last two actions, even when there is a visual input.  A quick way to fix this is to penalize it even more, but multiplying by two:

```dot(phrase, VERB*WRITE) - 2*dot(vision, WRITE+SAY+YES+NO+HI+BYE+OK)```

That seems to make it very reliable for me.